Fix null YAML values being replaced by `"null"`

matthewhughes934 opened the pull request:
Related issues:

* https://github.com/kubernetes-sigs/kustomize/issues/5031
* https://github.com/kubernetes-sigs/kustomize/issues/5171

After noting this behaviour was not present in
d89b448c745937f0cf1936162f26a5aac688f840 a `git bisect` pointed to the
change 1b7db20504ba8cb0ae4fa00037fe1676585b31a9. The issue with that
change is that upon seeing a `null` node it would replace it with a node
whose value was equivalent but without a `!!null` tag. This meant that
one application of a patch would have the desired approach: the result
would be `null` in the output, but on a second application of a similar
patch the field would be rendered as `"null"`.

To avoid this, define a new attribute on `RNode`s that is checked before
clearing any node we should keep. The added
`TestApplySmPatch_Idempotency` test verifies this behaviour.

See also https://github.com/kubernetes-sigs/kustomize/pull/5365 for an
alternative approach

justinsb commented on kyaml/yaml/fns.go:
```diff
@@ -725,7 +728,7 @@ func (s FieldSetter) Filter(rn *RNode) (*RNode, error) {
 	}
 
 	// Clear the field if it is empty, or explicitly null
```
I was wondering why this change was necessarily.  I tried it out and it was because otherwise the nil we so carefully preserved would be cleared.  Can we update the comment though, maybe adding "...unless it is our KeepNull marker" (though admittedly this makes the IsTaggedNull / "explicitly null" sort of confusing)

matthewhughes934 replied:
> I was wondering why this change was necessarily. I tried it out and it was because otherwise the nil we so carefully preserved would be cleared. Can we update the comment though, maybe adding "...unless it is our KeepNull marker" (though admittedly this makes the IsTaggedNull / "explicitly null" sort of confusing)

Good call, I tried to express the my understanding with [c18fcbc](https://github.com/kubernetes-sigs/kustomize/pull/5519/commits/c18fcbced6a098586dfc344ec248c1b88b897601)

justinsb commented:
I applied this patch locally and verified that it fixed the null -> "null" problem for me.  I also verified that if we don't have the `&& s.Value != KeepNullNode` that the nulls are removed entirely, so that line is needed (I didn't understand why at first, personally)

I don't think I have approval rights, but I may be able to mark this as...

/ok-to-test

stormqueen1990 commented:
/cc

matthewhughes934 commented:
An alternative (I think possibly nicer) approach that preserves the value of the `nil` field is to add a new attribute on RNodes like:

```diff
diff --git a/kyaml/yaml/fns.go b/kyaml/yaml/fns.go
index 77426ec38..68749b1fd 100644
--- a/kyaml/yaml/fns.go
+++ b/kyaml/yaml/fns.go
@@ -728,7 +728,7 @@ func (s FieldSetter) Filter(rn *RNode) (*RNode, error) {
        }
 
        // Clear the field if it is empty, or explicitly null
-       if (s.Value == nil || s.Value.IsTaggedNull()) && s.Value != KeepNullNode {
+       if s.Value == nil || (s.Value.IsTaggedNull() && !s.Value.PersistNull) {
                return rn.Pipe(Clear(s.Name))
        }
 
diff --git a/kyaml/yaml/merge2/merge2.go b/kyaml/yaml/merge2/merge2.go
index 4942092ad..c99e714e2 100644
--- a/kyaml/yaml/merge2/merge2.go
+++ b/kyaml/yaml/merge2/merge2.go
@@ -66,7 +66,9 @@ func (m Merger) VisitMap(nodes walk.Sources, s *openapi.ResourceSchema) (*yaml.R
 
                // If Origin is missing, preserve explicitly set null in Dest ("null", "~", etc)
                if nodes.Origin().IsNil() && !nodes.Dest().IsNil() && len(nodes.Dest().YNode().Value) > 0 {
-                       return yaml.KeepNullNode, nil
+                       n := yaml.NewScalarRNode(nodes.Dest().YNode().Value)
+                       n.PersistNull = true
+                       return n, nil
                }
 
                return nodes.Origin(), nil
diff --git a/kyaml/yaml/rnode.go b/kyaml/yaml/rnode.go
index bf3b5347f..2dee574d9 100644
--- a/kyaml/yaml/rnode.go
+++ b/kyaml/yaml/rnode.go
@@ -214,6 +214,8 @@ type RNode struct {
        // object root: object root
        value *yaml.Node
 
+       PersistNull bool
+
        Match []string
 }
```

stormqueen1990 reviewed (commented):
Hi there, @matthewhughes934! 👋🏻

Thank you very much for your contribution!

This is looking good.
However, could you please also add a test case that covers the current scenario (a null value being replaced by the word `"null"`), so we have some coverage to avoid regressions?

stormqueen1990 commented:
/label tide/merge-method-squash

matthewhughes934 commented:
> This is looking good. However, could you please also add a test case that covers the current scenario (a null value being replaced by the word `"null"`), so we have some coverage to avoid regressions?

Thanks for the review. The added test `TestApplySmPatch_Idempotency` functions as a regression test for this change.

just noting: I've updated things to use the approach I mentioned in a comment just above (because I found this was clearer), and squashed all the commits down

stormqueen1990 commented:
> > This is looking good. However, could you please also add a test case that covers the current scenario (a null value being replaced by the word `"null"`), so we have some coverage to avoid regressions?
> 
> Thanks for the review. The added test `TestApplySmPatch_Idempotency` functions as a regression test for this change.
> 
> just noting: I've updated things to use the approach I mentioned in a comment just above (because I found this was clearer), and squashed all the commits down

Hi there, @matthewhughes934! 👋🏻

I was looking for an explicit test that fails when the resulting YAML has the string `"null"`, but that's fair.

Thanks for squashing your commits!

stormqueen1990 commented:
/remove-label tide/merge-method-squash

stormqueen1990 reviewed (commented):
/lgtm

matthewhughes934 commented:
> I was looking for an explicit test that fails when the resulting YAML has the string `"null"`, but that's fair.

:+1: I've added a comment noting that it is the regression test for that issue (just squashed it into the latest change which I've also rebased on master)

stormqueen1990 reviewed (commented):
/lgtm

Thanks for adding the comment!

/cc @varshaprasad96
for maintainer review

varshaprasad96 reviewed (approved):
Thanks for the fix @matthewhughes934! The changes look good to me! 

/lgtm
/approve
